tests/unit/tf/horovod/test_horovod.py

```diff
@@ -0,0 +1,160 @@
+import os
+
+import pytest
+import tensorflow as tf
+from tensorflow.keras.utils import set_random_seed
+
+import merlin.models.tf as mm
+from merlin.datasets.advertising.criteo.dataset import default_criteo_transform
+from merlin.io.dataset import Dataset
+from merlin.models.tf.distributed.backend import hvd, hvd_installed
+from merlin.models.utils.example_utils import workflow_fit_transform
+from merlin.schema.tags import Tags
+
+# Set seed to make tests deterministic. In real training, it is not
+# necessary to set the random seed.
+set_random_seed(42)
+
+
+@pytest.mark.parametrize("custom_distributed_optimizer", [True, False])
+def test_horovod_multigpu_dlrm(
+    criteo_data,
+    tmpdir,
+    custom_distributed_optimizer,
+    batch_size=11,
+    learning_rate=0.03,
+):
+    """
+    This test should work on CPU, single GPU, and multiple GPUs.
+    However, for distributed training on multiple GPUs, it needs to be
+    executed with `horovodrun`:
+
+    $ horovodrun -np 2 hvd_wrapper.sh python -m pytest tests/unit/tf/horovod/test_horovod.py
+
+    Even with multiple GPUs, if you simply run pytest without horovodrun, it
+    will use only one GPU.
+    """
+    criteo_data.to_ddf().to_parquet(os.path.join(tmpdir, "train"))
+    criteo_data.to_ddf().to_parquet(os.path.join(tmpdir, "valid"))
+    workflow = default_criteo_transform()
+    workflow_fit_transform(
+        workflow,
+        os.path.join(tmpdir, "train", "*.parquet"),
+        os.path.join(tmpdir, "valid", "*.parquet"),
+        os.path.join(tmpdir, "processed"),
+    )
+
+    # As a workaround for nvtabular producing different numbers of batches in
+    # workers, we repartition the dataset in multiples of hvd.size().
+    # https://github.com/NVIDIA-Merlin/models/issues/765
+    train = Dataset(os.path.join(tmpdir, "processed", "train", "*.parquet"))
+    ddf = train.to_ddf().repartition(npartitions=2)
+    train = Dataset(ddf, schema=train.schema)
+
+    train_loader = mm.Loader(
```

Comment:
Does this loader need to have a global_size and global_rank supplied? (they can both be one I think...) we would want to test the multi-gpu part right? Same for next test.

Reply:
We set it [here](https://github.com/edknv/models/blob/6f9561059fa7bd060739641ab31baed21d3cd836/merlin/models/tf/loader.py#L299-L300) in the loader unless the user overrides it:
```python
            global_size = global_size or hvd.size()
            global_rank = global_rank or hvd.rank()
```
This PR is currently running tests on the Jenkins instance for the multi-part. I added a test run in [tox.ini](https://github.com/edknv/models/blob/6f9561059fa7bd060739641ab31baed21d3cd836/tox.ini#L33):
```
    horovodrun -np 2 sh examples/usecases/multi-gpu/hvd_wrapper.sh python -m pytest -m horovod -rxs tests/unit
```